Updating the user guide for MCSCF module (#35)
pyscf/pyscf.github.io: source/user/mcscf.rst

.. _user_mcscf:

Multi-configuration self-consistent field (MCSCF)
*************************************************

*Modules*: :mod:`mcscf`

Introduction
------------

Multiconfigurational self-consistent field (MCSCF) methods extend Hartree-Fock (HF) theory by describing the wave function as a linear combination of determinants. 

> Unlike full configuration interaction (FCI), described in :numref:`theory_ci`, complete active space (CAS) methods perform the FCI procedure on a subset of the molecular orbitals, referred to as the "active space." 

Reviewer: CASSCF is just one of countles MCSCF methods, I would write something like "There are many ways to construct a multiconfigurational  expansion, but the most widely used formulation are the complete active space (CAS) methods [https://doi.org/10.1016/0301-0104(80)80045-0] ..."

Reviewer 2: Sure, but the text is not wrong.

MCSCF methods other than CASSCF tend to be problematic in terms of multiple minima with respect to the orbitals.

These methods are crucial for systems that exhibit strong electron correlation, such as transition metal complexes.
For a detailed discussion of MCSCF methods, we direct the reader to :cite:`Helgaker2013` and :cite:`esqc`.

The MCSCF module has two main flavors: CASCI and CASSCF. 
CASCI uses the single reference orbitals from a HF calculation and solves the CI eigenvalue problem once.

> CASSCF calculations optimize the orbitals for the set of determinants and require solving the CI eigenvalue multiple (and possibly many) times.

Reviewer 2: The reference orbitals for CASCI may not be HF, you can also use other references like DFT, MP2 natural orbitals, etc.

Maybe it would be clearer to say that a CASSCF calculation consists of multiple CASCI steps, where the wave function computed at each step gives you an orbital gradient that is used to optimize the active space.

Author: Strictly speaking, I think CASCI orbitals should be HF, but practically speaking I know we can use other orbitals. I will change this tomorrow. I'll also remove the eigenvalue problem language as you suggested.

Reviewer: One can use any orbitals with CASCI. For example it is possible to run a small CAS and then use these orbitals in larger calculations, or obtain orbitals for one symmetry, and use it for another.

Reviewer 2: Yes. Common examples are CASCI(MP2) or CASCI(CISD). Using natural orbitals from MP2 or CI results in much faster convergence of the CI expansion than HF virtuals.

Minimal examples for each type of calculations are shown below.

CASCI
"""""
.. code-block:: python

  import pyscf
  mol = pyscf.M(
      atom = 'O 0 0 0; O 0 0 1.2',
      basis = 'ccpvdz',
      spin = 2)
  myhf = mol.RHF().run()
  ncas, nelecas = (6,8)

>   mycas = myhf.CASCI(ncas, nelecas).run()

Reviewer 2: Isn't this a bit of a funny example? How accurate are the HF orbitals?

Author: This is just the trimmed down version of this example:

https://github.com/pyscf/pyscf/blob/7be5e015b2b40181755c71d888449db936604660/examples/mcscf/00-simple_casci.py#L10-L20

Do you mean accurate in the sense that we'll get meaningful CASCI results?

Reviewer 2: Yes. IIRC HF virtuals usually give pretty poor estimates, which is why people have come up with a bunch of methods to get better virtual orbitals to match the HF occupied orbitals.

Author: Ok, that's a fair point, but I think the goal of these examples is to give people a minimum working example and favor ease of understanding rather than strict correctness (in terms of the results). That said, if people are strongly opposed to this, I'm happy to adjust the example.

Reviewer 2: I mean.. it would make a lot more sense if did the CAS-CI calculation in a basis of CISD natural orbitals :upside_down_face:

CASSCF
""""""
.. code-block:: python

  import pyscf
  mol = pyscf.M(
      atom = 'O 0 0 0; O 0 0 1.2',
      basis = 'ccpvdz',
      spin = 2)
  myhf = mol.RHF().run()

>   ncas, nelecas = (6,8)

Reviewer: Maybe it is good to show that one has to pass nelecas as a tuple to define spin?

Author: You're probably right, I can do that too

  mycas = myhf.CASSCF(ncas, nelecas).run()

Like many other modules in PySCF, :mod:`mcscf` works with density-fitting (:numref:`user_df`) and x2c (:numref:`user_x2c`).
It can also be used in its unrestricted formulation, please see :source:`examples/mcscf/60-uhf_based_ucasscf.py` for an example.

> An important feature of :mod:`mcscf` is that it can interface with external CI solver such as DMRG, FCIQMC, or selected CI methods, see the external projects for more details.

Reviewer 2: Or internal solvers like the coupled cluster or MP2 module!

Author: I can add this and link to the CC example.

Picking an Active Space
-----------------------
In general, selecting an active space can be cumbersome and PySCF offers severals ways to facilitate this process.

> There are four strategies and all require that you specify the number of electrons and orbitals in the active space.

Reviewer: Maybe it is good to explain that PySCF does a clever thing under the hood. The program actually uses the molecule charge to count electrons and add orbitals to active space from occupied or empty HF orbitals according to the input.

I also think it is important to distinguish the choice of the active space itself, which is fundamental from choice of starting orbitals that will be corrected by the SCF procedure.

Reviewer 2: 1) Yes. However, this is pretty obvious and is also used in other programs (at least Q-Chem does it). Also, this is how active spaces are always defined in the literature: (ne,nmo).

2) I don't agree: one cannot distinguish between the active space and the choice of the active orbitals. The orbitals **define** your active space. (E.g. if you have a diatomic molecule, the symmetry of the orbitals already defines your CAS problem: say, the 5s5p active space  = 10 sigma 5 pi, and this defines the Hamiltonian. You can always find the orbitals that minimize the CAS energy for this choice of the active space.)

3) "initial guess orbitals" is as correct as "starting orbitals". CAS-SCF is like SCF i.e. Hartree-Fock theory. You have an initial guess for the orbitals, and then you optimize them by minimizing the CAS-SCF energy.

Reviewer: >if you have a diatomic molecule, the symmetry of the orbitals already defines your CAS problem

The fact that the symmetry of inactive, active and virtual spaces is defined by starting orbitals is another 'smart' thing in PySCF. This actually should probably discussed as well.

In Molcas and Molpro to define CAS one has to give number of inactive and active orbitals per irrep together with number of active electrons. The common *(ne, nmo)* is just a short hand notation that implies a list of *nmo* per irrep.

.. warning::
  We always advise users to visualize their chosen active orbitals before starting large/expensive calculations.
  This involves dumping the MO coefficients to a ``molden`` file (see example :source:`examples/tools/02-molden.py`) and visualizing with your chosen program.
  While there are many great softwares available to visualize orbitals, `JMol <http://jmol.sourceforge.net/>`_ is one of the easiest to use and is recommended for less experienced users.

1) (Default) Specifying no additional information.

>   This is the most minimal strategy for selecting an active space and chooses orbitals (and electrons) around the fermi-level that match the number of orbitals and electrons specified by the user.

Reviewer 2: Fermi level

  In most circumstances, this is not an ideal strategy and will lead to poor convergence or none at all.
  For example:
.. code-block:: python

  ncas, nelecas = (6,8)
  mycas = myhf.CASSCF(ncas, nelecas)

> 

Reviewer 2: The traditional ways to pick active spaces are missing here. Often one runs an MP2 or CISD natural orbital calculation; and then picks the active space based on these natural orbitals. This leads to a much better guess than just HF orbitals.

Several other ways to pick the active space have also been suggested such as improved virtual orbitals (IVOs), which should be pretty easy to form in PySCF.

Author: You're right that I missed the natural orbitals, I can add those here. I was focusing strictly on methods where PySCF already contained the convenience functions for them in the mcscf modules. The natural orbitals are easy enough though so I can show that in a few lines.

Reviewer 2: Yes, please add an example to PySCF and `include` it here.

Reviewer: There is also the widely used strategy is to use just the minimal active that consists of either singly occupied orbitals or all transition-metal d orbitals.

Reviewer 2: @bogdanoff that's already covered by the AVAS example

However, what is also missing is Pulay's unrestricted HF natural orbitals, which should likewise be easy to implement in PySCF.

2) Specifying the molecular orbital (MO) index of the active space orbitals you want. 

>   The user can "manually" selects the MO orbital indices (in a 1-based indexing scheme) and passes them to the ``sort_mo`` function.

Reviewer 2: ```suggestion
The user can "manually" select the MO orbital indices (in a 1-based indexing scheme) and pass them to the ``sort_mo`` function.
```

  See :source:`examples/mcscf/10-define_cas_space.py` for more details.

.. code-block:: python

  mycas = mcscf.CASSCF(myhf, 4, 4)
  # Note sort_mo by default take the 1-based orbital indices.
  mo = mycas.sort_mo([5,6,8,9])
  mycas.kernel(mo)

> 

Reviewer 3: I would add here that it is useful to localize the orbitals when selecting the active space. Could you add that one may wish to localize the MOs and visualize them before selecting.

Reviewer 2: Exactly; picking the orbitals by index means very little when they are delocal.

> 3) Specifying the number of orbitals in each symmetry group. 

Reviewer: The symmetry composition of inactive, active and virtual spaces is fixed during the optimization. By default it is defied by starting orbitals.

Author: @bogdanoff, is this something you want added or are you saying you don't think strategy 3) is a useful one?

Reviewer 3: I am going to edit this PR to resolve these comments, let's freeze further comments.

Author: I'm about to push my work from today so you may want to wait for 10 or so minutes

This strategy can occasionally be helpful when the initial guess orbitals are not easily identifiable.

.. code-block:: python

  mycas = mcscf.CASSCF(mf, 12, 12)
  ncore = {'A1g':5, 'A1u':5}
  ncas = {'A1g':2, 'A1u':2,'E1ux':1, 'E1uy':1, 'E1gx':1, 'E1gy':1,
              'E2ux':1, 'E2uy':1, 'E2gx':1, 'E2gy':1}
  mo = mcscf.sort_mo_by_irrep(mycas, mf.mo_coeff, ncas, ncore)
  mycas.kernel(mo)

A similar approach where we specify the electron occupations by irreducible representation is also possible by setting ``mycas.fcisolver.irrep_nelec``.

.. code-block:: python

  mycas = mcscf.CASSCF(myhf, 8, 8)
  mycas.fcisolver.irrep_nelec = {"A1g": (2, 1), "A1u": (1, 1), "E1ux": (1, 1), "E1uy": (1, 0)}

4) Use automated strategies (``avas`` and ``dmet_cas``) to pick an active space based on AO orbitals you're targeting.
For more details, see :source:`examples/mcscf/43-avas.py` and :source:`examples/mcscf/43-dmet_cas.py`.

.. code-block:: python

  from pyscf.mcscf import avas
  ao_labels = ['Fe 3d', 'Fe 4d', 'C 2pz']
  ncas, nelecas, orbs = avas.avas(mf, ao_labels)
  mycas = mcscf.CASSCF(mf, ncas, nelecas)

>   

Reviewer 3: Is dmet-cas and AVAS just the same thing?

Reviewer 2: I think AVAS actually runs localization.

.. code-block:: python

  from pyscf.mcscf import dmet_cas
  ao_labels = ['Fe 3d', 'Fe 4d', 'C 2pz']
  ncas, nelecas, mo = dmet_cas.guess_cas(mf, mf.make_rdm1(), ao_labels)
  mycas = mcscf.CASSCF(mf, ncas, nelecas)
  mycas.kernel(mo)

Frozen Core MCSCF
-----------------

To reduce to computational expense of CASSCF calculations, users can "freeze" orbitals thereby excluding them from optimization.

> Users can specify a number to free the N lowest orbitals:

Reviewer 2: ```suggestion
Users can specify the number of lowest orbitals to freeze:
```

.. code-block:: python

  mycas = mcscf.CASSCF(myhf, 6, 8)
  mycas.frozen = 2
  mycas.kernel()

Users can also specify a list of orbital indices (0-based).

> These may be occupied, virtual, or active orbitals.

Reviewer 2: Active orbitals are by definition not frozen!

Author: While I agree that active orbitals should never really be frozen, I found that they can be when looking at the `examples/19-frozen_core.py`, see below:

https://github.com/pyscf/pyscf/blob/7be5e015b2b40181755c71d888449db936604660/examples/mcscf/19-frozen_core.py#L36-L40

If people think it will just confuse users, I can say that only occupied or virtual orbitals can be frozen even though that's not strictly true.

Reviewer 2: No, the point is that that description is wrong. If you assign the orbitals to the frozen core, then they are frozen even though they would be otherwise chosen as active.

I think it's a good point to make that the code is flexible enough to do this: it might be that your system has some orbitals close to the Fermi level that aren't really entangled, so freezing them saves a lot of time.

Reviewer 2: So the correct expression is that one can also freeze orbitals that are either occupied or virtual in the reference determinant that are close to the Fermi level.

Reviewer: Frozen means not optimized during SCF. So I think active orbitals can be frozen, i.e. not optimized, yet their occupation will not be restricted in the wave-function expansion, i.e. they are still active.

Reviewer 2: The `frozen` argument controls the **frozen core**, which *is* included in the orbital optimization; however, the occupation is restricted to two electrons for occupied orbitals, and zero electrons for virtuals.

Reviewer: Exactly, `frozen` keyword controls restrictions of orbital optimization, but has nothing to do with occupation restrictions, which are controlled by active space definition. The keyword is smart in PySCF, if you give a number, it will define a **frozen core**, but if you give a list, it will exclude from orbital optimization orbital from any space.

Reviewer 2: @bogdanoff no, the exact opposite: https://sunqm.github.io/pyscf/ci.html
```
frozen
: int or list

    If integer is given, the inner-most orbitals are frozen from CI amplitudes. 
```

**Frozen core** means the orbitals do not participate in CI calculations. It is exactly used to define the active space. For instance, if you freeze the 1s orbitals in H2O, you're left with 8 electrons. For a planar system like a polyacene, one typically freezes the sigma electrons, and correlates only the pi electrons. This is the use case where you want to set the frozen orbitals as a list.

Reviewer 2: This is an easy point of confusion, since different codes have different definitions. For instance, in Psi4 you can 1) freeze the core with the `freeze_core` setting, **or** 2) set the orbitals separately, where they split to a) FROZEN orbitals, b) restricted orbitals, c) active orbitals, d) inactive orbitals, e) frozen virtuals [you can also define the active and inactive orbitals with RAS spaces]. Frozen orbitals are either doubly occupied or not unoccupied, and they are not optimized. Restricted and inactive orbitals are also either doubly occupied or unoccupied; however, they *are* included in the orbital optimization.

So, the difference is between frozen core, and frozen orbital.

Reviewer 2: This is also telling us that the difference between frozen core and frozen orbital needs to be discussed thoroughly in the documentation.

Feel free to copy-paste my explanation above.

Reviewer 2: Aaaandd... I now see your point, @bogdanoff: this is indeed an argument for the optimizer :see_no_evil:

... which means that I am still right, only in the totally opposite sense :rofl:

Reviewer: Indeed, one has to use of word *frozen* with care.
In Molpro, there is a `freeze, orb.sym` keyword in mcscf module, which prevents orbital `orb.sym` from optimization. But no word *frozen* is directly used for active orbitals.
Just for the record text from  [https://sunqm.github.io/pyscf/mcscf.html](https://sunqm.github.io/pyscf/mcscf.html):
```
frozen
: int or list
    If integer is given, the inner-most orbitals are excluded from optimization.
    Given the orbital indices (0-based) in a list, any doubly occupied core orbitals,
    active orbitals and external orbitals can be frozen.
```
Probably the most important use of freezing active orbitals is x-ray spectroscopy with hole in deep core orbital that otherwise will be optimized away from the active space.

.. code-block:: python
  mycas = mcscf.CASSCF(myhf, 6, 8)
  mycas.frozen = [0,1,26,27]
  mycas.kernel()

See :source:`examples/mcscf/19-frozen_core.py` for a complete example.

State-Averaged Calculations
---------------------------

> When dealing with states that are close in energy, it can be helpful to perform state average calculations where the multireference orbitals are optimized for multiple states.

Reviewer 2: I'm not sure whether it's suffciently clear that "states" mean "CAS wave functions".

There are no "multireference" orbitals; orbitals are always one-electron states.

Reviewer: State-averaging is also needed to compute excited-states energies.  During SCF procedure all orbitals are optimized, not only active.

Author: @susilehtola in this context "multireference orbitals" refers to the fact that the orbitals are not optimal for a single determinant, but multiple determinants. To make that clearer, I can add a similar explanation in the docs.

Reviewer 2: @jamesETsmith yes, but "orbitals" don't mean they've been optimized in any way :)

Reviewer 2: But yes, a better term would be CAS orbitals

The ``state_average_`` function (note the hanging underscore) is a member function of ``CASCI``/``CASSCF`` objects and takes the weights of the states as input.

> The weights can be any normalized and non-negative array of values.

Reviewer: Nevertheless,most widely state averaging with equal weights is used.

Reviewer 2: ```suggestion
The weights can be any normalized and non-negative array of values.
Equal weights are often, but not always, used.
 ```

Reviewer 2: What if the weights given in are not normalized?

Say, if the user inputs an array `[1 1 2 3 5 8]`, does the program use these weights as is, normalize them, or what?

Reviewer 2: I also wonder why negative weights are not allowed.

Author: I don't think negative values wouldn't make sense when you calculate the energy using these weights see the docs here https://github.com/pyscf/pyscf/blob/7be5e015b2b40181755c71d888449db936604660/pyscf/mcscf/addons.py#L788

If you chose all negative weights your energy would flip sign.

There is an assertion check see below:
https://github.com/pyscf/pyscf/blob/7be5e015b2b40181755c71d888449db936604660/pyscf/mcscf/addons.py#L808

.. code-block:: python

  n_states = 5
  weights = np.ones(n_states)/n_states
  mycas = mcscf.CASSCF(mf, 4, 4).state_average_(weights)

See :source:`examples/mcscf/15-state_average.py` for a complete example.

By default, only a single spin and/or point group symmetry is targeted, but it is possible to target a mixture of both:

.. code-block:: python

  weights = [.5, .5]
  solver1 = fci.direct_spin1_symm.FCI(mol)
  solver1.wfnsym= 'A1'
  solver1.spin = 0
  solver2 = fci.direct_spin1_symm.FCI(mol)
  solver2.wfnsym= 'A2'
  solver2.spin = 2

  mycas = mcscf.CASSCF(mf, 4, 4)
  mcscf.state_average_mix_(mycas, [solver1, solver2], weights)
  mycas.kernel()

See :source:`examples/mcscf/41-state_average.py` for a complete example.

Job Control
-----------

Optimization Settings
"""""""""""""""""""""

> For CASSCF calculations, users may want to modify several of the convergence tolerances such as the energy tolerance (``conv_tol``), the orbital gradient tolerance (``conv_tol_grad``), and the maximum number of MCSCF iteration (``max_cycle_macro``).

Reviewer 2: Replace "energy tolerance" and "orbital gradient tolerance" with convergence thresholds for the energy and the orbital gradient

.. code-block:: python

  mycas = mcscf.CASSCF(mf, 6, 6)
  mycas.conv_tol = 1e-12
  mycas.conv_tol_grad = 1e-6
  mycas.max_cycle_macro = 25
  mycas.kernel()

Initial Guess
"""""""""""""

> Initial guess orbitals for the CASSCF calculation may be passed to the ``kernel`` member function of an MCSCF object.

Reviewer: Starting orbitals

Reviewer 2: No, this is fine. We use the same language as in the SCF section.

We can say
```suggestion
Initial guess orbitals for the CASSCF calculation (starting orbitals) may be passed to the ``kernel`` member function of an MCSCF object.
```

.. code-block:: python

  mycas = mcscf.CASSCF(myhf, 8, 8)
  mycas.kernel(my_custom_mos)

> CI coefficient from previous or related calculations can also be passed as an initial guess to expedite a calculation:

Reviewer 2: ```suggestion
CI coefficients from a previous calculation can also be passed as an initial guess to expedite the calculation:
```

Reviewer 2: Then one can note that the state does not have to be the same

.. code-block:: python
  mycas = mcscf.CASSCF(myhf, 8, 8)
  mycas.kernel(my_custom_mos, my_custom_ci)

Examples:

* :source:`examples/mcscf/14-project_init_guess.py`
* :source:`examples/mcscf/31-cr2_scan/cr2-scan.py`
* :source:`examples/mcscf/34-init_guess_localization.py`
* :source:`examples/mcscf/43-avas.py`
* :source:`examples/mcscf/43-dmet_cas.py`

Restarting
""""""""""

.. warning::
  When running large calculations, it's always recommended that you specify a checkpoint file for your calculation.

.. code-block:: python

  mycas.chkfile = "casscf.chk"

> Much like :mod:`scf`, if a job is interrupted, users can restart the MCSCF calculations using checkpoint files from a crashed calculation.

Reviewer: Another use is to start from smaller active space orbitals or different spin/symmetry sector.

Reviewer 2: @bogdanoff great addition!

.. code-block:: python

  from pyscf.lib import chkfile
  old_chk_file = "old_casscf.chk"
  mycas = mcscf.CASSCF(scf.RHF(mol), 6, 6)
  mycas.chkfile = "restarted_casscf.chk"
  mo = chkfile.load(old_chk_file, 'mcscf/mo_coeff')
  mycas.kernel(mo)

See :source:`examples/mcscf/13-restart.py` for a complete example.

Observables and Properties
--------------------------

Wave Function Analysis
""""""""""""""""""""""

The ``analyze`` member functions of MCSCF objects prints many useful properties to ``stdout`` when the verbosity is >=4.

1) Natural orbital occupancies
2) Natural orbital AO expansions
3) Overlap between canonical MCSCF orbitals and the initial guess orbitals.
4) Analysis of the CI coefficients, i.e. the leading configurations and their weights
5) AO populations
6) Atomic populations
7) AO spin densities (if applicable)
8) Atomic spin densities (if applicable)

.. code-block:: python
  mycas = myhf.CASCI(6, 8).run()
  mycas.verbose = 4
  mycas.analyze()

Natural Orbitals
""""""""""""""""

> Users can request that the converged MCSCF orbitals be transformed with the member attribute ``natorb``

Reviewer: Energy of CAS state is invariant under orbital rotation within inactive, active, and virtual sectors.
Inactive and virtual orbitals are by default canonicalized, i.e. they transformed such that Fock matrices within virtual and inactive sectors are diagonal, and orbital energy can be assigned to these orbitals.
By default active orbitals are kept untouched after orbital optimization and strictly speaking no energy or electron occupation can be assigned to them.
Users can request that active orbitals be transformed to the so-called natural representation, such that the one-body density matrix is diagonal and electron occupation can be assigned to them.

Author: Thanks for this @bogdanoff, I'll add this to the docs tonight.

.. warning::

>   When ``mycas.natorb`` is set, the natural orbitals may NOT be sorted by the active space occupancy.

Reviewer 2: I wonder why this is; it sounds like the code should always sort the active space orbitals in occupation so that the active space window can be modified without problems...

.. _user_mcscf:

Multi-configuration self-consistent field (MCSCF)
*************************************************

*Modules*: :mod:`mcscf`

Introduction
------------

> Multiconfigurational self-consistent field (MCSCF) methods extend Hartree-Fock (HF) theory by describing the wave function as a linear combination of determinants. 

Reviewer 2: ```suggestion
Multiconfigurational self-consistent field (MCSCF) methods go beyond the single-determinant Hartree-Fock (HF) method by allowing the wave function to become a linear combination of multiple determinants.
```

> While there are a large variety of MCSCF methods, PySCF focuses on the complete active space (CAS) family of methods.
> Unlike full configuration interaction (FCI), described in :numref:`theory_ci`, CAS methods perform the FCI procedure on a subset of the molecular orbitals, referred to as the "active space." 

Reviewer 2: ```suggestion
While the configurations i.e. determinants can in principle be chosen in an arbitrary number of ways, PySCF focuses on the complete active space (CAS) family of methods, where the set of electron configurations is defined in terms of a given set of active orbitals, also known as the "active space".
The CAS method generates all possible electron configurations that can be formed from the set of the active orbitals, and is therefore equivalent to an FCI procedure on a subset of the molecular orbitals; please see :numref:`theory_ci` for a discussion on the FCI method.
```

> These methods are crucial for systems that exhibit strong electron correlation, such as transition metal complexes.

Reviewer 2: ```suggestion
The use of MCSCF methods is crucial for reliable modeling of systems that exhibit nearly degenerate orbitals i.e. static correlation, such as transition metal complexes.
```

For a detailed discussion of MCSCF methods, we direct the reader to References :cite:`Helgaker2013` and :cite:`esqc`.

The MCSCF module has two main flavors: CASCI and CASSCF. 

> CASCI expands the wave function as a linear combination of slater determinants and variationally solves for the coefficients of this expansion.
> CASSCF calculations optimize the orbitals for the set of determinants and the coefficients of their expansion.
> This means performing multiple (and possibly many) CASCI step.

Reviewer 2: ```suggestion
In CASCI, the wave function is written as a linear combination of Slater determinants, and the expansion coefficients are solved in a variational procedure.
The orbitals are fixed in a CASCI calculation.
In contrast, in a CASSCF calculation, the orbitals are relaxed to minimize the resulting CASCI energy.
This means performing multiple CASCI steps followed by updates to the molecular orbitals, in analogy to Hartree-Fock calculations which also relies on iterative algorithms to minimize the HF energy with respect to the orbitals.
```

Minimal examples for each type of calculations are shown below.

CASCI
"""""
.. code-block:: python

  from pyscf import gto, mp, mcscf
  mol = gto.M(
      atom = 'O 0 0 0; O 0 0 1.2',
      basis = 'ccpvdz',
      spin = 2)
  myhf = mol.RHF().run()

>   mymp = mp.UMP2(myhf).run()

Reviewer 2: ```suggestion
  # Use MP2 natural orbitals to define the active space for the single-point CAS-CI calculation
  mymp = mp.UMP2(myhf).run()
```

In CASCI, the wave function is written as a linear combination of Slater determinants, and the expansion coefficients are solved in a variational procedure.
The orbitals are fixed in a CASCI calculation.
In contrast, in a CASSCF calculation, the orbitals are relaxed to minimize the resulting CASCI energy.
This means performing multiple CASCI steps followed by updates to the molecular orbitals, in analogy to Hartree-Fock calculations which also relies on iterative algorithms to minimize the HF energy with respect to the orbitals.

Minimal examples for each type of calculations are shown below.

CASCI
"""""
.. code-block:: python

  from pyscf import gto, mp, mcscf
  mol = gto.M(
      atom = 'O 0 0 0; O 0 0 1.2',
      basis = 'ccpvdz',
      spin = 2)
  myhf = mol.RHF().run()
  mymp = mp.UMP2(myhf).run()

  noons, natorbs = mcscf.addons.make_natural_orbitals(mymp)
  ncas, nelecas = (6,8)
  mycas = mcscf.CASCI(myhf, ncas, nelecas)
  mycas.kernel(natorbs)

CASSCF
""""""
.. code-block:: python

  import pyscf
  mol = pyscf.M(
      atom = 'O 0 0 0; O 0 0 1.2',
      basis = 'ccpvdz',
      spin = 2)
  myhf = mol.RHF().run()
  ncas, nelecas = (6,(5,3))

>   mycas = myhf.CASSCF(ncas, nelecas).run()

Reviewer 2: ```suggestion
   # We can also run CAS calculations starting from the Hartree-Fock orbitals.
   mycas = myhf.CASSCF(ncas, nelecas).run()
```

It can also be used in its unrestricted formulation, please see :source:`examples/mcscf/60-uhf_based_ucasscf.py` for an example.
An important feature of :mod:`mcscf` is that it can interface with external CI solver such as DMRG, FCIQMC, or selected CI methods, see the external projects for more details.
You can even use MP2 or CC methods as solvers!
Check out :source:`/examples/cc/42-as_casci_fcisolver.py` for a working example.

Picking an Active Space
-----------------------
In general, selecting an active space can be cumbersome and PySCF offers severals ways to facilitate this process.
There are several strategies and they all contain two main components:

* Specify the number of electrons and orbitals in the active space.
* (Optional) Specify the which orbitals are "active" 

.. 
  warning::
  The total set orbitals (core, active, and virtual) used in active space methods can be specified or selected in a variety of ways, giving users substantial flexibility for CAS-type calculations.
  But users should note, "with great power comes great responsibility."

>   Active space calculations are notoriously difficult and just because a calculation completes without error does not guarantee that the results will be chemically/physically meaningful so we urge users to select their active space orbitals with thought and care.

Reviewer 2: ```suggestion
  Active space calculations are notoriously difficult and just because a calculation completes without error does not guarantee that the results will be chemically/physically meaningful.
  We urge users to select their active space orbitals with thought and care.
  You should always try out several choices for the active space!
```

.. note::
  We always advise users to visualize their chosen active orbitals before starting large/expensive calculations.
  This involves dumping the MO coefficients to a ``molden`` file (see example :source:`examples/tools/02-molden.py`) and visualizing with your chosen program.
  While there are many great softwares available to visualize orbitals, `JMol <http://jmol.sourceforge.net/>`_ is one of the easiest to use and is recommended for less experienced users.

Below is a list of several general strategies one could employ to pick active space orbitals:

1) (Default) Specifying no additional information.
  This is the most minimal strategy for selecting an active space and chooses orbitals (and electrons) around the Fermi level that match the number of orbitals and electrons specified by the user.
  In most circumstances, this is not an ideal strategy and will lead to poor convergence or none at all.

  For example:
.. code-block:: python

  ncas, nelecas = (6,8)
  mycas = myhf.CASSCF(ncas, nelecas)

> 

Reviewer 2: ```suggestion

.. note::
  Hartree-Fock orbitals are often poor for systems with significant static correlation.
  In such cases, orbitals from density functional calculations often yield better starting points for CAS calculations.
```

  We urge users to select their active space orbitals with thought and care.
  You should always try out several choices for the active space!

.. note::
  We always advise users to visualize their chosen active orbitals before starting large/expensive calculations.
  This involves dumping the MO coefficients to a ``molden`` file (see example :source:`examples/tools/02-molden.py`) and visualizing with your chosen program.
  While there are many great softwares available to visualize orbitals, `JMol <http://jmol.sourceforge.net/>`_ is one of the easiest to use and is recommended for less experienced users.

Below is a list of several general strategies one could employ to pick active space orbitals:

1) (Default) Specifying no additional information.
  This is the most minimal strategy for selecting an active space and chooses orbitals (and electrons) around the Fermi level that match the number of orbitals and electrons specified by the user.
  In most circumstances, this is not an ideal strategy and will lead to poor convergence or none at all.

  For example:
.. code-block:: python

  ncas, nelecas = (6,8)
  mycas = myhf.CASSCF(ncas, nelecas)

> 2) Specifying the molecular orbital (MO) index of the active space orbitals you want. 

Reviewer 2: ```suggestion
 2) Specifying the active space orbitals as a list of molecular orbital (MO) indices. 
```

  While there are many great softwares available to visualize orbitals, `JMol <http://jmol.sourceforge.net/>`_ is one of the easiest to use and is recommended for less experienced users.

Below is a list of several general strategies one could employ to pick active space orbitals:

1) (Default) Specifying no additional information.
  This is the most minimal strategy for selecting an active space and chooses orbitals (and electrons) around the Fermi level that match the number of orbitals and electrons specified by the user.
  In most circumstances, this is not an ideal strategy and will lead to poor convergence or none at all.

  For example:
.. code-block:: python

  ncas, nelecas = (6,8)
  mycas = myhf.CASSCF(ncas, nelecas)

.. note::
  Hartree-Fock orbitals are often poor for systems with significant static correlation.
  In such cases, orbitals from density functional calculations often yield better starting points for CAS calculations.

2) Specifying the molecular orbital (MO) index of the active space orbitals you want. 

>   This is often useful after selecting (and typically visualizing) localized orbitals.

Reviewer 2: ```suggestion
   This is often useful combined with a visual analysis of localized orbitals (see the section on localized orbitals).
```

>   The user can "manually" select the MO orbital indices (in a 1-based indexing scheme) and pass them to the ``sort_mo`` function.

Reviewer 2: ```suggestion
  The user can select the MO orbital indices "manually" and pass them to the ``sort_mo`` function. NB! The orbitals are numbered from 1, not 0.
```

  See :source:`examples/mcscf/10-define_cas_space.py` and :source:`examples/mcscf/34-init_guess_localization.py` for more details.

.. code-block:: python

  mycas = mcscf.CASSCF(myhf, 4, 4)
  # Note sort_mo by default take the 1-based orbital indices.
  mo = mycas.sort_mo([5,6,8,9])
  mycas.kernel(mo)

3) Specifying the number of orbitals in each symmetry group. 

> This strategy can occasionally be helpful when the initial guess orbitals are not easily identifiable.

Reviewer 2: ```suggestion
This strategy is often helpful when the system has a high degree of symmetry, or when there is no physically evident best choise for the active orbitals.
```

.. code-block:: python

  mycas = mcscf.CASSCF(mf, 12, 12)
  ncore = {'A1g':5, 'A1u':5}
  ncas = {'A1g':2, 'A1u':2,'E1ux':1, 'E1uy':1, 'E1gx':1, 'E1gy':1,
              'E2ux':1, 'E2uy':1, 'E2gx':1, 'E2gy':1}
  mo = mcscf.sort_mo_by_irrep(mycas, mf.mo_coeff, ncas, ncore)
  mycas.kernel(mo)

> A similar approach where we specify the electron occupations by irreducible representation is also possible by setting ``mycas.fcisolver.irrep_nelec``.

Reviewer 2: ```suggestion
A similar approach where we specify the number of electrons in each irreducible representation is also possible by setting ``mycas.fcisolver.irrep_nelec``.
```

.. code-block:: python

  mycas = mcscf.CASSCF(myhf, 8, 8)
  mycas.fcisolver.irrep_nelec = {"A1g": (2, 1), "A1u": (1, 1), "E1ux": (1, 1), "E1uy": (1, 0)}

> 

Reviewer 2: ```suggestion
.. note::
  This strategy is often combined with calculations at a lower level of theory.
  For instance, MP2 or CISD natural orbitals and their occupation numbers can be used determine the suitable active space in each symmetry block.
  Natural orbitals with occupations close to 2 are strongly occupied, and can be frozen in the CAS calculation.
  Natural orbitals with small occupation numbers can likewise be omitted from the active space.
```

4) Use automated strategies (``avas`` and ``dmet_cas``) to pick an active space based on AO orbitals you're targeting.
For more details, see :source:`examples/mcscf/43-avas.py` and :source:`examples/mcscf/43-dmet_cas.py`.

.. code-block:: python

  from pyscf.mcscf import avas
  ao_labels = ['Fe 3d', 'Fe 4d', 'C 2pz']
  ncas, nelecas, orbs = avas.avas(mf, ao_labels)
  mycas = mcscf.CASSCF(mf, ncas, nelecas)

  

.. code-block:: python

  from pyscf.mcscf import dmet_cas
  ao_labels = ['Fe 3d', 'Fe 4d', 'C 2pz']
  ncas, nelecas, mo = dmet_cas.guess_cas(mf, mf.make_rdm1(), ao_labels)
  mycas = mcscf.CASSCF(mf, ncas, nelecas)
  mycas.kernel(mo)

> Frozen Core MCSCF

Reviewer 2: ```suggestion
 Frozen-orbital MCSCF
```

  mycas.fcisolver.irrep_nelec = {"A1g": (2, 1), "A1u": (1, 1), "E1ux": (1, 1), "E1uy": (1, 0)}
.. note::
  This strategy is often combined with calculations at a lower level of theory.
  For instance, MP2 or CISD natural orbitals and their occupation numbers can be used determine the suitable active space in each symmetry block.
  Natural orbitals with occupations close to 2 are strongly occupied, and can be frozen in the CAS calculation.
  Natural orbitals with small occupation numbers can likewise be omitted from the active space.
4) Use automated strategies (``avas`` and ``dmet_cas``) to pick an active space based on AO orbitals you're targeting.
For more details, see :source:`examples/mcscf/43-avas.py` and :source:`examples/mcscf/43-dmet_cas.py`.

.. code-block:: python

  from pyscf.mcscf import avas
  ao_labels = ['Fe 3d', 'Fe 4d', 'C 2pz']
  ncas, nelecas, orbs = avas.avas(mf, ao_labels)
  mycas = mcscf.CASSCF(mf, ncas, nelecas)

  

.. code-block:: python

  from pyscf.mcscf import dmet_cas
  ao_labels = ['Fe 3d', 'Fe 4d', 'C 2pz']
  ncas, nelecas, mo = dmet_cas.guess_cas(mf, mf.make_rdm1(), ao_labels)
  mycas = mcscf.CASSCF(mf, ncas, nelecas)
  mycas.kernel(mo)

>  Frozen-orbital MCSCF

Reviewer 2: ```suggestion
Frozen-orbital MCSCF
```

-----------------

> To reduce to computational expense of CASSCF calculations, users can "freeze" orbitals thereby excluding them from optimization.

Reviewer 2: ```suggestion
Orbitals can be frozen in the orbital optimization to e.g. reduce the computational effort of CASSCF calculations.
The orbitals will remain fixed throughout the optimization.
```

For more details, see :source:`examples/mcscf/43-avas.py` and :source:`examples/mcscf/43-dmet_cas.py`.

.. code-block:: python

  from pyscf.mcscf import avas
  ao_labels = ['Fe 3d', 'Fe 4d', 'C 2pz']
  ncas, nelecas, orbs = avas.avas(mf, ao_labels)
  mycas = mcscf.CASSCF(mf, ncas, nelecas)

  

.. code-block:: python

  from pyscf.mcscf import dmet_cas
  ao_labels = ['Fe 3d', 'Fe 4d', 'C 2pz']
  ncas, nelecas, mo = dmet_cas.guess_cas(mf, mf.make_rdm1(), ao_labels)
  mycas = mcscf.CASSCF(mf, ncas, nelecas)
  mycas.kernel(mo)

Frozen-orbital MCSCF
-----------------

Orbitals can be frozen in the orbital optimization to e.g. reduce the computational effort of CASSCF calculations.
The orbitals will remain fixed throughout the optimization.

Users can specify the number of lowest orbitals to freeze:

.. code-block:: python

  mycas = mcscf.CASSCF(myhf, 6, 8)
  mycas.frozen = 2
  mycas.kernel()

Users can also specify a list of orbital indices (0-based).
These may be occupied, virtual, or active orbitals.

.. code-block:: python
  mycas = mcscf.CASSCF(myhf, 6, 8)
  mycas.frozen = [0,1,26,27]
  mycas.kernel()

> See :source:`examples/mcscf/19-frozen_core.py` for a complete example.

Reviewer 2: ```suggestion
See :source:`examples/mcscf/19-frozen_core.py` for a complete example.

.. note::
  The `frozen` keyword of the CASSCF optimizer should not be confused with the `frozen` keyword of the FCI solver, which controls the number of orbitals that are constrained to be doubly occupied.
```

State-Averaged Calculations
---------------------------

> When dealing with states that are close in energy, it can be helpful to perform state average calculations where the orbitals are optimized for multiple states.

Reviewer 2: ```suggestion
When your system has states that are close-by in energy, their ordering may change during the orbital optimization and result in non-convergence of the CASSCF optimization.
In such cases it is often helpful to optimize the orbitals for a state average.
```

The orbitals will remain fixed throughout the optimization.

Users can specify the number of lowest orbitals to freeze:

.. code-block:: python

  mycas = mcscf.CASSCF(myhf, 6, 8)
  mycas.frozen = 2
  mycas.kernel()

Users can also specify a list of orbital indices (0-based).
These may be occupied, virtual, or active orbitals.

.. code-block:: python
  mycas = mcscf.CASSCF(myhf, 6, 8)
  mycas.frozen = [0,1,26,27]
  mycas.kernel()

See :source:`examples/mcscf/19-frozen_core.py` for a complete example.

.. note::
  The `frozen` keyword of the CASSCF optimizer should not be confused with the `frozen` keyword of the FCI solver, which controls the number of orbitals that are constrained to be doubly occupied.

State-Averaged Calculations
---------------------------

When your system has states that are close-by in energy, their ordering may change during the orbital optimization and result in non-convergence of the CASSCF optimization.
In such cases it is often helpful to optimize the orbitals for a state average.
The ``state_average_`` function (note the hanging underscore) is a member function of ``CASCI``/``CASSCF`` objects and takes the weights of the states as input.

> The weights can be any normalized and non-negative array of values, but typically they are all the same.

Reviewer 2: ```suggestion
The weights can be any normalized and non-negative array of values, but equal weights are typically used.
```

.. code-block:: python

  from pyscf.lib import chkfile
  old_chk_file = "old_casscf.chk"
  mycas = mcscf.CASSCF(scf.RHF(mol), 6, 6)
  mycas.chkfile = "restarted_casscf.chk"
  mo = chkfile.load(old_chk_file, 'mcscf/mo_coeff')
  mycas.kernel(mo)

See :source:`examples/mcscf/13-restart.py` for a complete example.

Restarting calculations can be also be useful when using results from a smaller active space to speed up calculations on a larger one.

Observables and Properties
--------------------------

Wave Function Analysis
""""""""""""""""""""""

The ``analyze`` member functions of MCSCF objects prints many useful properties to ``stdout`` when the verbosity is >=4.

1) Natural orbital occupancies
2) Natural orbital AO expansions
3) Overlap between canonical MCSCF orbitals and the initial guess orbitals.
4) Analysis of the CI coefficients, i.e. the leading configurations and their weights
5) AO populations
6) Atomic populations
7) AO spin densities (if applicable)
8) Atomic spin densities (if applicable)

.. code-block:: python

  mycas = myhf.CASCI(6, 8).run()
  mycas.verbose = 4
  mycas.analyze()

Natural Orbitals
""""""""""""""""

> Energy of CAS state is invariant under orbital rotation within inactive, active, and virtual sectors.
> Inactive and virtual orbitals are by default canonicalized, i.e. they transformed such that Fock matrices within virtual and inactive sectors are diagonal, and orbital energy can be assigned to these orbitals.
> By default active orbitals are kept untouched after orbital optimization and strictly speaking no energy or electron occupation can be assigned to them.
> Users can request that active orbitals be transformed to the so-called natural representation, such that the one-body density matrix is diagonal and electron occupation can be assigned to them.

Reviewer 2: ```suggestion
The CAS wave function is invariant under orbital rotation within the inactive core, active, and inactive virtual spaces.
The inactive core and virtual orbitals are by default canonicalized, i.e., they are transformed by block-diagonalizing the generalized Fock matrix within the inactive core and virtual subblocks.
This also yields energy eigenvalues for the inactive core and virtual orbitals.
In contrast, no energy or electron occupation can be assigned to the active orbitals, strictly speaking.
The user may request the active orbitals to be transformed to the so-called natural representation, in which the reduced one-body density matrix is diagonal, and the formal natural orbital occupation numbers can be found on its diagonal.
```
